Add a content parameter to aws_s3 put operation

onitake opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
This is a port of the original pull request at https://github.com/ansible/ansible/pull/67091 to collections. There might still be compatibility issues due to renamed packages.

Add support for uploading S3 objects from in-memory content in addition to on-disk files.

Fixes https://github.com/ansible/ansible/issues/67086
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
aws_s3

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
Example usage:
<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below
- name: PUT operation from a rendered template
  aws_s3:
    bucket: mybucket
    object: /object.yaml
    content: "{{ lookup('template', 'templates/object.yaml.j2') }}"
    mode: put
```

onitake commented:
Comment from original PR:

> Hmm... Looks like strings aren't autocast into byte strings.
> If possible, I'd like to keep the possibility to pass a binary data into the content parameter. Is this possible at all? Should I change the content parameter type to string?

onitake commented:
Another comment:

> So, I added a unit test for the content parameter and it looks like the feature's working as expected. Shippable had a timeout in a completely unrelated test, perhaps a retry is in order.
>
> I changed the `content` parameter type to string, as I don't see a way to pass byte strings into Ansible modules. If there's a way to do this, please tell me. The string is converted to UTF-8 before sending it to S3.
>
> What the unit tests don't cover yet is a multi-part put. This is probably crucial, as the the new `calculate_etag_content` utility function will not have full test coverage otherwise. I'll try to address this.

tremble commented on tests/integration/targets/aws_s3/tasks/main.yml:
```diff
@@ -562,6 +562,60 @@
               - result is not changed
       when: ansible_system == 'Linux' or ansible_distribution == 'MacOSX'
 
+    - name: create an object from static content
```
It would be good to test idempotency here:
Run the same 'put' again and see if changed==False

onitake replied:
Makes sense.

onitake commented on plugins/modules/aws_s3.py:
```diff
@@ -666,13 +694,14 @@ def main():
         dualstack=dict(default='no', type='bool'),
         rgw=dict(default='no', type='bool'),
         src=dict(),
```
Oh, cool. Didn't know about that.

onitake commented on plugins/modules/aws_s3.py:
```diff
@@ -792,8 +822,10 @@ def main():
-            module.fail_json(msg="Local object for PUT does not exist")
+        if content is not None and src is not None:
+            module.fail_json(msg="Either src or content must be specified, but not both")
+        if content is None and not path_check(src):
```
Makes sense, I just adapted this from the base code. Together with the `mutually_exclusive` above, this should cover all cases. Is `path_check` even necessary when we do `type='path'`?

tremble replied:
`type='path'` validates that it follows the standard path syntax and does things like expanding '~', it doesn't validate that the path exists, as you might also be creating the path.

onitake replied:
I also noticed that the same may apply for `dest` further up. There is only a `path_check(dest)`, but `dest` is not a required parameter. Should I add a None check there as well?

onitake commented on plugins/modules/aws_s3.py:
```diff
@@ -792,8 +822,10 @@ def main():
 
-        if not path_check(src):
-            module.fail_json(msg="Local object for PUT does not exist")
+        if content is not None and src is not None:
```
Nice, didn't know about that.

onitake commented:
I also noticed that there a ton of % formats in the module.
Since Python 2 is now EOL, wouldn't it also be good to change them into f-strings or at least str.format()?

onitake commented:
I addressed your comments and also fixed a bug with etag calculation that popped up thanks to the idempotency test.

My other two comments regarding `dest` checking and refactoring % into `format()` have nothing to do with this PR, so please disregard them.

tremble commented:
Short verison on `%` : Yes we probably should (although I don't always remember)

The only trouble with testing that dest exists is that you need to check that the parent folder exists since you may be creating the target.  (but it would be worth tweaking that to type=path too)

tremble commented on plugins/modules/aws_s3.py:
```diff
@@ -131,6 +131,15 @@
+      - This parameter will be treated as a string and converted to UTF-8 before sending it to S3.
+        To send binary data, use the I(src) parameter instead.
+      - Either I(content) or I(src) must be specified for a PUT operation. Ignored otherwise.
+    version_added: "2.10"
```
@jillr What are we supposed to be using for 'version_added' at the minute?

jillr replied:
@tremble the collection version the patch will land in, using semver.  We should be targeting `version_added: 1.1.0` right now.

jillr commented:
Regarding python2, Ansible will support 2.7 as long as we have supported platforms that ship with 2.7 (like RHEL).  So it would be good to update these but it's not urgent.

jillr commented:
@onitake following up on your question:
> If possible, I'd like to keep the possibility to pass a binary data into the content parameter. Is this possible at all? Should I change the content parameter type to string?

Unfortunately there's not any good ways to take multiple types of content in the same parameter.  The best way to handle this would be to have a `content` parameter that is for `str` data, and another option such as `base64_content` for `bytes` data.

abadger commented:
And the `base64_content` parameter would take a base64 encoded byte string (not a raw byte string).  There's no way in ansible to pass a raw byte string through all the layers.

onitake commented:
Thanks for the suggestions, I will add such a parameter.

onitake commented:
I implemented the mentioned `content_base64` parameter, but now we we have a bit of asymmetry:
There is no `getbin` mode to obtain binary object contents without writing to disk.
I think this is an acceptable tradeoff, but if you want it added as well, I'd rather submit a separate PR for that feature.

tremble commented on plugins/modules/aws_s3.py:
```diff
@@ -126,6 +126,22 @@
+      - The parameter value will be treated as a string and converted to UTF-8 before sending it to S3.
+        To send binary data, use the I(content_base64) parameter instead.
+      - Either I(content), I(content_base64) or I(src) must be specified for a PUT operation. Ignored otherwise.
+    version_added: "1.1.0"
```
```suggestion
    version_added: "1.3.0"
```

tremble commented on plugins/modules/aws_s3.py:
```diff
@@ -126,6 +126,22 @@
+      - The base64-encoded binary data to PUT into an object.
+      - Use this if you need to put raw binary data, and don't forget to encode in base64.
+      - Either I(content), I(content_base64) or I(src) must be specified for a PUT operation. Ignored otherwise.
+    version_added: "1.1.0"
```
```suggestion
    version_added: "1.3.0"
```

tremble commented:
Many thanks for your efforts on this.  I'm sorry it's taken so long.

Given how long you've been waiting on us for this I've added a changelog fragment and tweaked the 'added_in' value.

onitake commented:
Thanks for taking time to review this & merging it!
